packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/undo/handlers/UndoEventHandler.java

```diff
@@ -0,0 +1,73 @@
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+
+import org.eclipse.emf.common.notify.Notification;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.sirius.components.collaborative.api.ChangeDescription;
+import org.eclipse.sirius.components.collaborative.api.ChangeKind;
+import org.eclipse.sirius.components.collaborative.api.IEditingContextEventHandler;
+import org.eclipse.sirius.components.core.api.ErrorPayload;
+import org.eclipse.sirius.components.core.api.IEditingContext;
+import org.eclipse.sirius.components.core.api.IInput;
+import org.eclipse.sirius.components.core.api.IPayload;
+import org.eclipse.sirius.components.core.api.SuccessPayload;
+import org.eclipse.sirius.web.application.editingcontext.EditingContext;
+import org.eclipse.sirius.web.application.undo.dto.UndoInput;
+import org.springframework.stereotype.Service;
+
+import reactor.core.publisher.Sinks.Many;
+import reactor.core.publisher.Sinks.One;
+
+/**
+ * Handler used to undo mutations.
+ *
+ * @author mcharfadi
+ */
+@Service
+public class UndoEventHandler implements IEditingContextEventHandler {
+
+    @Override
+    public boolean canHandle(IEditingContext editingContext, IInput input) {
+        return input instanceof UndoInput;
+    }
+
+    @Override
+    public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDescriptionSink, IEditingContext editingContext, IInput input) {
+        IPayload payload = new ErrorPayload(input.id(), "Error ");
+        ChangeDescription changeDescription = new ChangeDescription(ChangeKind.SEMANTIC_CHANGE, editingContext.getId(), input);
+        if (editingContext instanceof EditingContext siriusEditingContext && input instanceof UndoInput undoInput) {
+            var changes = siriusEditingContext.getChanges();
+
+            var change = changes.get(undoInput.mutationId());
+            //Testing with an int value from a form textfield
+            for (Notification notification : change) {
+                if (notification.getOldValue() != null) {
+                    if (notification.getNotifier() instanceof EObject eObject) {
+                        EStructuralFeature feature = (EStructuralFeature) notification.getFeature();
+
+                        //This line is throwing java.util.ConcurrentModificationException
+                        //I tried wrapping the block of code before the for loop in a synchronised(changes) {...}
+                        // but if I do I still get the error and the feature is no longer set
+                        eObject.eSet(feature, notification.getOldIntValue());
```

Comment:
You are iterating on the changes of the resources while simultaneously creating new changes in this very same collection by modifying the resource set :)

You can create an issue like that very easily:

```
for (var item: list) {
  list.add(new Object);
}
```

The iterator used by the for will throw the exception because you are modifying it while iterating on it.

Reply:
I tried to use a `synchronizedList` before but it did not work, I don't know the implication of using `CopyOnWriteArrayList`, I'll look into that

Reply:
You're correct that if I stop listening to the changes made while restoring, it fixes the issue too.